source/applications/dependency-injection.md

```diff
@@ -0,0 +1,183 @@
+You can create your own factory type by simply registering a factory with the
+new type. For example, to create a `user` type, you'd simply register your
+factory with `application.register('user:user-to-register')`.
+
+Factory registrations must be performed either in application or application
+instance initializers (with the former being much more common).
+
+For example, an application initializer could register a `Logger` factory with
+the key `logger:main`:
+
+```app/initializers/logger.js
+export function initialize(application) {
+  var Logger = Ember.Object.extend({
+    log(m) {
+      console.log(m);
+    }
+  });
+
+  application.register('logger:main', Logger);
+}
+```
+
+### Registering Already Instantiated Objects
+
+By default, Ember will attempt to instantiate a registered factory when it is
+looked up. When registering an already instantiated object instead of a class,
+use the `instantiate: false` option to avoid attempts to re-instantiate it
+during lookups.
+
+In the following example, the `logger` is a plain JavaScript object that should
+be returned "as is" when it's looked up:
+
+```app/initializers/logger.js
+export function initialize(application) {
+  var logger = {
+    log(m) {
+      console.log(m);
+    }
+  };
+
+  application.register('logger:main', logger, { instantiate: false });
+}
+```
+
+### Registering Singletons vs. Non-Singletons
+
+By default, registrations are treated as "singletons". This simply means that
+an instance will be created when it is first looked up, and this same instance
+will be cached and returned from subsequent lookups.
+
+When you want fresh objects to be created for every lookup, register your
+factories as non-singletons using the `singleton: false` option.
+
+In the following example, the `Message` class is registered as a non-singleton:
+
+```app/initializers/logger.js
+export function initialize(application) {
+  var Message = Ember.Object.extend({
+    text: ''
+  });
+
+  application.register('notification:message', Message, { singleton: false });
+}
+```
+
+## Factory Injections
+
+Once a factory is registered, it can be "injected" where it is needed.
+
+Factories can be injected into whole "types" of factories with *type
+injections*. For example:
+
+```app/initializers/logger.js
+export function initialize(application) {
+  var Logger = Ember.Object.extend({
+    log(m) {
+      console.log(m);
+    }
+  });
+
+  application.register('logger:main', Logger);
+  application.inject('route', 'logger', 'logger:main');
+}
+```
+
+As a result of this type injection, all factories of the type `route` will be
+instantiated with the property `logger` injected. The value of `logger` will
+come from the factory named `logger:main`.
+
+Routes in this example application can now access the injected logger:
+
+```app/routes/index.js
+export default Ember.Route.extend({
+  activate() {
+    // The logger property is injected into all routes
+    this.logger.log('Entered the index route!');
```

Comment:
Though I know it isn't required, I would prefer we continue to suggest `this.get('logger')` here as it is more refactoring-friendly (works with injected services).